src/components/TreeGrid.ts

```diff
@@ -0,0 +1,220 @@
+import _ from 'lodash/fp';
+import { ReactElement, useState } from 'react';
+import { div, h, strong } from 'react-hyperscript-helpers';
+import { Grid } from 'react-virtualized';
+import { Link } from 'src/components/common/Link';
+import { icon } from 'src/components/icons';
+import colors from 'src/libs/colors';
+import { DEFAULT, switchCase, toIndexPairs } from 'src/libs/utils';
+
+export type RowContents = {
+  // The id could be generated by `wrapContent` if we can't rely on the data having an id field.
+  /** The row's unique identifier. */
+  id: number;
+  /** If true, this row has children. */
+  hasChildren: boolean;
+};
+
+export type Column<T extends RowContents> = {
+  /** The column's name, displayed as a column header. */
+  name: string;
+  /** The column's width, in pixels. */
+  width: number;
+  /** Given the current row, return the column's contents. */
+  render: (row: T) => string | ReactElement;
+};
+
+type RowState = 'closed' | 'opening' | 'open';
+
+type Row<T extends RowContents> = {
+  contents: T;
+  depth: number;
+  isFetched: boolean;
+  state: RowState;
+};
+
+/**
+ *  Wrap a RowContents in a Row object.
+ *  @param depth the tree depth of the wrapped row
+ */
+const wrapContent =
+  (depth: number) =>
+  <T extends RowContents>(contents: T): Row<T> => ({
+    contents,
+    depth,
+    isFetched: false,
+    state: 'closed',
+  });
+
+type TreeGridProps<T extends RowContents> = {
+  /** the columns to display */
+  readonly columns: Column<T>[];
+  /** the initial rows to display */
+  readonly initialRows: T[];
+  /** Given a row, return its children. This is only called if row.hasChildren is true. */
+  readonly getChildren: (row: T) => Promise<T[]>;
+};
+
+type TreeGridPropsInner<T extends RowContents> = TreeGridProps<T> & {
+  readonly gridWidth: number;
+};
+
+/**
+ * Given all rows, return the rows that are currently visible.
+ */
+function getVisibleRows<T extends RowContents>(allRows: Row<T>[]) {
+  const visibleRows: Row<T>[] = [];
+  // Root elements are always visible.
+  let visibleDepth = 0;
+  for (const row of allRows) {
+    // A row is not visible if it's deeper than the current deepest visible row.
+    if (row.depth > visibleDepth) {
+      continue;
+    }
+    if (row.state === 'open') {
+      // If a row is open, then all rows deeper than it are also visible.
+      visibleDepth = row.depth + 1;
+    } else {
+      // If a row is closed, the visible depth is now the current row's depth.
+      visibleDepth = row.depth;
+    }
+    visibleRows.push(row);
+  }
+  return visibleRows;
+}
+
+const getRowIndex = <T extends RowContents>(row: Row<T>, rows: Row<T>[]) =>
+  _.findIndex((r) => r.contents.id === row.contents.id, rows);
```

Comment:
Isn't the array supposed to go before the predicate?  e.g.
```suggestion
  _.findIndex(rows, (r) => r.contents.id === row.contents.id);
```

Reply:
Well, terra-ui uses lodash-fp, not lodash, so the args are in a different order then you'd expect. Collections are passed last and all function support a curried forms (e.g., `_.findIndex((r) => ...)` produces a function that accepts `rows`. So this could be written as
```
const getRowIndex = <T extends RowContents>(row: Row<T>) =>
  _.findIndex((r: Row<T>) => r.contents.id === row.contents.id);
```

But then uses would also have to be curried, like this:
```
    const index = getRowIndex(row)(data);
```

See https://github.com/lodash/lodash/wiki/FP-Guide